README.rst
MiraTitanHMFemulator
============================

This package provides the emulator for the halo mass function (HMF) from the
*Mira-Titan Universe* suite of cosmological *N*-body simulations.

Installation
------------

This python package only requires the standard packages numpy and scipy. Simply
install using ``pip``::

  pip install MiraTitanHMFemulator

and then ``pytest`` to validate your installation::

  cd /path/to/MiraTitanHMFemulator
  pytest

With every test passed you are good to go!

Documentation
-------------

Documentation is hosted at `ReadTheDocs
<http://MiraTitanHMFemulator.readthedocs.io/>`_. Find examples for how to use
the emulator there!

Citation
--------

Please cite our paper when you use the *Mira-Titan* HMF emulator::

  @ARTICLE{Bocquet2020ApJ...901....5B,
           author = {{Bocquet}, Sebastian and {Heitmann}, Katrin and {Habib}, Salman and
                     {Lawrence}, Earl and {Uram}, Thomas and {Frontiere}, Nicholas and
                     {Pope}, Adrian and {Finkel}, Hal},
           title = "{The Mira-Titan Universe. III. Emulation of the Halo Mass Function}",
           journal = {\apj},
           keywords = {Large-scale structure of the universe, 902, Astrophysics - Cosmology and Nongalactic Astrophysics},
           year = 2020,
           month = sep,
           volume = {901},
           number = {1},
           eid = {5},
           pages = {5},
           doi = {10.3847/1538-4357/abac5c},
           archivePrefix = {arXiv},
           eprint = {2003.12116},
           primaryClass = {astro-ph.CO},
           adsurl = {https://ui.adsabs.harvard.edu/abs/2020ApJ...901....5B},
           adsnote = {Provided by the SAO/NASA Astrophysics Data System}}


Feedback, problems, questions
-----------------------------

Please `report issues on GitHub
<https://github.com/SebastianBocquet/MiraTitanHMFemulator/issues>`_ where the
package is hosted.

.. image:: https://zenodo.org/badge/225578117.svg
   :target: https://zenodo.org/badge/latestdoi/225578117


CHANGELOG.rst
Version 0.1.1
  * python2 now also produces correct results

Version 0.1.0
  * Do not check parameter limits for wa since the parameter hypercube is defined in w0-wb space.

Version 0.0.1
  * Allow user to provide cosmology parameters without underscores

Version 0.0.0
  * This is the first public release. It went live with the arXiv submission.


tests/test_MiraTitanHMFemulator.py
import numpy as np
import inspect
import os
import pytest

import MiraTitanHMFemulator

class TestClass:
    z_arr = np.linspace(0, 2.02, 4)
    m_arr = np.logspace(13, 16, 31)

    def test_init(self):
        self.HMFemu = MiraTitanHMFemulator.Emulator()


    def test_validated_params(self):
        HMFemu = MiraTitanHMFemulator.Emulator()
        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        assert HMFemu.validate_params(fiducial_cosmo) is True

        # Missing keys, parameter limits
        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo.pop(k)
            assert HMFemu.validate_params(_cosmo) is False

            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]+= 2
            assert HMFemu.validate_params(_cosmo) is False

            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]-= 2
            assert HMFemu.validate_params(_cosmo) is False


    def test_missingkey(self):
        HMFemu = MiraTitanHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo.pop(k)
            with pytest.raises(KeyError):
                HMFemu.predict(_cosmo, self.z_arr, self.m_arr)

        with pytest.raises(TypeError):
            HMFemu.predict()


    def test_cosmo_limits(self):
        HMFemu = MiraTitanHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }

        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]+= 2
            with pytest.raises(ValueError):
                HMFemu.predict(_cosmo, self.z_arr, self.m_arr)

        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]-= 2
            with pytest.raises(ValueError):
                HMFemu.predict(_cosmo, self.z_arr, self.m_arr)


    def test_fiducial(self):
        np.random.seed(1328)
        data_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))

        HMFemu = MiraTitanHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }

        res = HMFemu.predict(fiducial_cosmo, self.z_arr, self.m_arr)

        # Mass function
        _fname = os.path.join(data_path, 'fid.npy')
        # np.save(_fname, res[0])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[0], ref))

        # Error on mass function
        _fname = os.path.join(data_path, 'fid_err.npy')
        # np.save(_fname, res[1])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[1], ref))


    def test_center(self):
        np.random.seed(1328)
        data_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))

        HMFemu = MiraTitanHMFemulator.Emulator()

        mid_cosmo = {}
        for k in ['Ommh2', 'Ombh2', 'Omnuh2', 'n_s', 'h', 'sigma_8', 'w_0', 'w_a']:
            mid_cosmo[k] = .5 * np.sum(HMFemu.param_limits[k])

        res = HMFemu.predict(mid_cosmo, self.z_arr, self.m_arr)

        # Mass function
        _fname = os.path.join(data_path, 'mid.npy')
        # np.save(_fname, res[0])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[0], ref))

        # Error on mass function
        _fname = os.path.join(data_path, 'mid_err.npy')
        # np.save(_fname, res[1])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[1], ref))


tests/test_MiraTitanUniverseHMFemulator.py
import numpy as np
import inspect
import os
import pytest

import MiraTitanUniverseHMFemulator

class TestClass:

    def test_init(self):
        self.HMFemu = MiraTitanUniverseHMFemulator.Emulator()


    def test_validated_params(self):
        HMFemu = MiraTitanUniverseHMFemulator.Emulator()
        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        assert HMFemu.validate_params(fiducial_cosmo) is True

        # Missing keys, parameter limits
        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo.pop(k)
            assert HMFemu.validate_params(_cosmo) is False

            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]+= 2
            assert HMFemu.validate_params(_cosmo) is False

            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]-= 2
            assert HMFemu.validate_params(_cosmo) is False


    def test_missingkey(self):
        HMFemu = MiraTitanUniverseHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo.pop(k)
            with pytest.raises(KeyError):
                HMFemu.predict(_cosmo)

        with pytest.raises(TypeError):
            HMFemu.predict()


    def test_limits(self):
        HMFemu = MiraTitanUniverseHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }

        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]+= 2
            with pytest.raises(ValueError):
                HMFemu.predict(_cosmo)

        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]-= 2
            with pytest.raises(ValueError):
                HMFemu.predict(_cosmo)


    def test_fiducial(self):
        data_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))
        z_arr = [2.02, 1.61, 1.01, 0.656, 0.434, 0.242, 0.101, 0.0]

        HMFemu = MiraTitanUniverseHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }

        res = HMFemu.predict(fiducial_cosmo)

        for i in range(len(z_arr)):
            _fname = os.path.join(data_path, 'fid_%d.npy'%i)
            # np.save(_fname, res[z_arr[i]]['HMF'])
            ref = np.load(_fname)
            assert np.all(np.isclose(ref, res[z_arr[i]]['HMF']))


    def test_center(self):
        data_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))
        z_arr = [2.02, 1.61, 1.01, 0.656, 0.434, 0.242, 0.101, 0.0]

        HMFemu = MiraTitanUniverseHMFemulator.Emulator()

        mid_cosmo = {}
        for k in ['Ommh2', 'Ombh2', 'Omnuh2', 'n_s', 'h', 'sigma_8', 'w_0', 'w_a']:
            mid_cosmo[k] = .5 * np.sum(HMFemu.param_limits[k])

        res = HMFemu.predict(mid_cosmo)

        for i in range(len(z_arr)):
            _fname = os.path.join(data_path, 'mid_%d.npy'%i)
            # np.save(_fname, res[z_arr[i]]['HMF'])
            ref = np.load(_fname)
            assert np.all(np.isclose(ref, res[z_arr[i]]['HMF']))


MiraTitanHMFemulator/tests/test_MiraTitanHMFemulator.py
import numpy as np
import inspect
import os
import pytest

import MiraTitanHMFemulator

class TestClass:
    z_arr = np.linspace(0, 2.02, 4)
    m_arr = np.logspace(13, 16, 31)


    def test_init(self):
        self.HMFemu = MiraTitanHMFemulator.Emulator()


    def test_translate_params(self):
        HMFemu = MiraTitanHMFemulator.Emulator()
        fiducial_cosmo_no_underscore = {'Ommh2': .3*.7**2,
                                        'Ombh2': .022,
                                        'Omnuh2': .006,
                                        'ns': .96,
                                        'h': .7,
                                        'w0': -1,
                                        'wa': 0,
                                        'sigma8': .8}

        # Check that translate_params returns
        assert HMFemu._Emulator__translate_params(fiducial_cosmo_no_underscore) is True

        # Check that translate_params copied the parameters
        for var_w, var_wo in zip(['w_0', 'w_a', 'n_s', 'sigma_8'], ['w0', 'wa', 'ns', 'sigma8']):
            assert np.isclose(fiducial_cosmo_no_underscore[var_wo], fiducial_cosmo_no_underscore[var_w])

        # Check that inconsistent definitions are caught
        variables_with_underscores = {'n_s': 1,
                                      'w_0': -1.1,
                                      'w_a': .1,
                                      'sigma_8': .88}
        for name in variables_with_underscores.keys():
            _cosmo = fiducial_cosmo_no_underscore.copy()
            _cosmo[name] = variables_with_underscores[name]
            assert HMFemu._Emulator__translate_params(_cosmo) is False


    def test_validate_params(self):
        HMFemu = MiraTitanHMFemulator.Emulator()
        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        assert HMFemu.validate_params(fiducial_cosmo) is True

        fiducial_cosmo_no_underscore = {'Ommh2': .3*.7**2,
                                        'Ombh2': .022,
                                        'Omnuh2': .006,
                                        'ns': .96,
                                        'h': .7,
                                        'w0': -1,
                                        'wa': 0,
                                        'sigma8': .8,
                                        }
        assert HMFemu.validate_params(fiducial_cosmo_no_underscore) is True

        # Missing keys, parameter limits
        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo.pop(k)
            assert HMFemu.validate_params(_cosmo) is False

            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]+= 2
            assert HMFemu.validate_params(_cosmo) is False

            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]-= 2
            assert HMFemu.validate_params(_cosmo) is False


    def test_missingkey(self):
        HMFemu = MiraTitanHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }
        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo.pop(k)
            with pytest.raises(KeyError):
                HMFemu.predict(_cosmo, self.z_arr, self.m_arr)

        with pytest.raises(TypeError):
            HMFemu.predict()


    def test_cosmo_limits(self):
        HMFemu = MiraTitanHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }

        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]+= 2
            with pytest.raises(ValueError):
                HMFemu.predict(_cosmo, self.z_arr, self.m_arr)

        for k in fiducial_cosmo.keys():
            _cosmo = fiducial_cosmo.copy()
            _cosmo[k]-= 2
            with pytest.raises(ValueError):
                HMFemu.predict(_cosmo, self.z_arr, self.m_arr)


    def test_fiducial(self):
        np.random.seed(1328)
        data_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))

        HMFemu = MiraTitanHMFemulator.Emulator()

        fiducial_cosmo = {'Ommh2': .3*.7**2,
                          'Ombh2': .022,
                          'Omnuh2': .006,
                          'n_s': .96,
                          'h': .7,
                          'w_0': -1,
                          'w_a': 0,
                          'sigma_8': .8,
                          }

        res = HMFemu.predict(fiducial_cosmo, self.z_arr, self.m_arr)

        # Mass function
        _fname = os.path.join(data_path, 'fid.npy')
        # np.save(_fname, res[0])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[0], ref))

        # Error on mass function
        _fname = os.path.join(data_path, 'fid_err.npy')
        # np.save(_fname, res[1])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[1], ref))


    def test_center(self):
        np.random.seed(1328)
        data_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))

        HMFemu = MiraTitanHMFemulator.Emulator()

        mid_cosmo = {}
        for k in ['Ommh2', 'Ombh2', 'Omnuh2', 'n_s', 'h', 'sigma_8', 'w_0']:
            mid_cosmo[k] = .5 * np.sum(HMFemu.param_limits[k])
        mid_cosmo['w_a'] = .5 * (-1.73 + 1.28)

        res = HMFemu.predict(mid_cosmo, self.z_arr, self.m_arr)

        # Mass function
        _fname = os.path.join(data_path, 'mid.npy')
        # np.save(_fname, res[0])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[0], ref))

        # Error on mass function
        _fname = os.path.join(data_path, 'mid_err.npy')
        # np.save(_fname, res[1])
        ref = np.load(_fname)
        assert np.all(np.isclose(res[1], ref))


MiraTitanHMFemulator/GP_matrix.py
import numpy as np
from scipy import linalg

class GaussianProcess:
    def compute_rho_corr_func_point(self, a, b, this_rho):
        """Compute correlation between two points a and b."""
        corr = np.prod(this_rho**(4 * (a - b)**2))
        return corr


    def compute_rho_corr_func(self, a, b, this_rho):
        """Compute rho correlation function between two vectors a and b.
        Returns kernel matrix [len(a), len(b)]."""
        corr_matrix = [self.compute_rho_corr_func_point(a[i], b[j], this_rho)
            for i in range(a.shape[0]) for j in range(b.shape[0])]
        return np.array(corr_matrix).reshape(a.shape[0], -1)


    def __init__(self, x, y, cov_n, prec_f, rho, compute_lnlike=False):
        """Set up covariance matrix and pre-compute its Cholesky decomposition.
        Parameters
        ----------
            x: design points [N_data, N_dim_input]
            y: design values [N_data, N_output]
            cov_n: covariance of y [N_output*N_data, N_output*N_data]
            prec_f: precision of the GP
            rho: CP correlation length [N_output, N_dim_input]
            compute_lnlike: (optional, default to False) marginal likelihood
        Returns
        -------
            None
        """
        self.N_data = len(x)
        self.N_dim_input = x.shape[1]
        self.N_output = y.shape[1]

        # Check dimensions
        if self.N_data!=len(y):
            raise TypeError("len(design points) %d must match len(design values) %d"%(self.N_data, len(y)))
        if len(prec_f)!=self.N_output:
            raise TypeError("len(prec_f) %d must match number of outputs %d"%(len(prec_f), self.N_output))
        if cov_n.shape!=(self.N_output*self.N_data, self.N_output*self.N_data):
            raise TypeError("Shape of data cov mat (%s,%s) must be (%s,%s)."%(
                cov_n.shape[0], cov_n.shape[1], self.N_output*self.N_data, self.N_output*self.N_data))
        if rho.shape!=(self.N_output, self.N_dim_input):
            raise TypeError("Shape of correlation lengths (%d,%d) must be (%d,%d)"%(
                rho.shape[0], rho.shape[1], self.N_output*self.N_dim_input, self.N_output*self.N_dim_input))

        self.x = x
        self.corr_rho = rho
        self.prec_f = prec_f
        self.y_flat = y.flatten(order='F')

        # Correlation matrix
        self.corrmat = np.zeros((self.N_output*self.N_data, self.N_output*self.N_data))
        for i in range(self.N_output):
            self.corrmat[i*self.N_data:(i+1)*self.N_data, i*self.N_data:(i+1)*self.N_data] = self.compute_rho_corr_func(x, x, self.corr_rho[i])/self.prec_f[i]
        try:
            self.cholesky_factor = linalg.cho_factor(self.corrmat + cov_n)
        except:
            print("Could not compute Cholesky decomposition")
            return
        self.Krig_basis = linalg.cho_solve(self.cholesky_factor, self.y_flat)

        if compute_lnlike:
            chi_squared = np.matmul(self.y_flat.T, self.Krig_basis)
            ln_corrmat_det = 2 * np.sum(np.log(np.diag(self.cholesky_factor[0])))
            self.lnlike = -.5 * chi_squared - .5 * ln_corrmat_det


    def predict(self, x_new):
        """
        Parameters: evaluation points [N_dim_input]
        Returns: (mean, variance)
        """

        if len(x_new)!=self.N_dim_input:
            raise TypeError("Evaluation points %s needs to be shape %d"%(len(x_new), self.N_dim_input))

        # Correlation with design input [N_output, N_data]
        corr_xnew_x = np.zeros((self.N_output, self.N_output*self.N_data))
        for i in range(self.N_output):
            corr_xnew_x[i,i*self.N_data:(i+1)*self.N_data] = [self.compute_rho_corr_func_point(x_new, self.x[j], self.corr_rho[i])
                                                              for j in range(self.N_data)]
        corr_xnew_x/= self.prec_f[:,None]

        # Mean prediction
        eval_mean = np.dot(corr_xnew_x, self.Krig_basis)

        # Variance
        v = linalg.cho_solve(self.cholesky_factor, corr_xnew_x.T)
        eval_covmat = np.diag(1./self.prec_f) - np.dot(corr_xnew_x, v)

        return eval_mean, eval_covmat


MiraTitanUniverseHMFemulator/MiraTitanUniverseHMFemulator.py
import numpy as np
import os
import inspect

from . import GP_matrix as GP


class Emulator:
    """The Mira-Titan Universe emulator for the halo mass function.

    Attributes
    -----------------
    param_limits : dictionary
        Lower and upper limits of the cosmological parameters.
    z_arr : array
        Redshifts of the emulator output.
    """
    # Cosmology parameters
    __param_names = ('Ommh2', 'Ombh2', 'Omnuh2', 'n_s', 'h', 'sigma_8', 'w_0', 'w_b')
    param_limits = {
        'Ommh2': (.12, .155),
        'Ombh2': (.0215, .0235),
        'Omnuh2': (0, .01),
        'n_s': (.85, 1.05),
        'h': (.55, .85),
        'sigma_8': (.7, .9),
        'w_0': (-1.3, -.7),
        'w_a': (-1.73, 1.28),
        'w_b': (.3, 1.3),
    }
    # Emulator redshifts
    z_arr = [2.02, 1.61, 1.01, 0.656, 0.434, 0.242, 0.101, 0.0]



    def __init__(self):
        """No arguments are required when initializing an `Emulator` instance.
        Upon initialization, a bunch of matrix operations are performed which
        takes a few seconds."""
        data_path = os.path.join(os.path.dirname(os.path.abspath(inspect.stack()[0][1])), 'data')

        # Basis functions and PCA standardization parameters
        # They have different lengths so they are stored in separate files
        self.__PCA_means, self.__PCA_transform = [], []
        for i in range(len(self.z_arr)):
            filename = os.path.join(data_path, 'PCA_mean_std_transform_%d.npy'%i)
            _tmp = np.load(filename)
            self.__PCA_means.append(_tmp[0,:])
            self.__PCA_transform.append(_tmp[1:,:])
        self.__GP_means = np.load(os.path.join(data_path, 'GP_params_mean.npy'))
        self.__GP_std = np.load(os.path.join(data_path, 'GP_params_std.npy'))

        # GP input data
        params_design = np.load(os.path.join(data_path, 'params_design_w0wb.npy'))
        hyper_params = np.load(os.path.join(data_path, 'hyperparams.npy'))
        input_means = np.load(os.path.join(data_path, 'means.npy'))
        cov_mat_data = np.load(os.path.join(data_path, 'cov_n.npy'))
        N_PC = input_means.shape[2]
        self.__GPreg = [GP.GaussianProcess(params_design,
                                           input_means[z_id],
                                           cov_mat_data[z_id],
                                           hyper_params[z_id,:N_PC],
                                           hyper_params[z_id,N_PC:].reshape(N_PC,-1))
                        for z_id in range(len(self.z_arr))]



    def predict(self, requested_cosmology, N_draw=0, return_draws=False):
        """Emulates the halo mass function dN/dlnM for the desired set of
        cosmology parameters and returns an output dictionary.

        Parameters
        ----------
        requested_cosmology : dictionary
            The set of cosmology parameters for which the HMF is requested. The
            parameters are `Ommh2`, `Ombh2`, `Omnuh2`, `n_s`, `h`,
            `sigma_8`, `w_0`, `w_a`.
        N_draw : int, optional
            The emulator can provide error estimates for its HMF prediction. If
            `N_draw` is provided, the emulator also provides `N_draw`
            samples of the HMF and the standard deviation of those estimates.
        return_draws : bool, optional
            If True, also return the HMF draws from the emulator posterior
            distribution. Default is False to save memory. Applies only if
            `N_draw` is > 0.

        Returns
        -------
        output : dictionary
            A dictionary containing all the emulator output. A `readme` key
            describes the units: The HMFs are dN/dlnM [(h/Mpc)^3]. The output is
            organized by redshift -- each dictionary key corresponds to one
            redshift. For each redshift, the output contains the following data:

        redshift : float
            The redshift of the emulator output.

        log10_M : array
            Decadic logarithm log10(mass [Msun/h]) for which the HMF is
            emulated.

        HMF : array
            The HMF corresponding to the mean emulated parameters.

        HMF_mean : array, optional
            The HMF corresponding to the mean of the HMFs drawn from the
            emulator posterior distribution. Only if `N_draw` is > 0.

        HMF_std : array, optional
            The (relative) standard deviation in the HMF draws from the emulator
            posterior distribution. Should be used if as a relative error on the
            HMF. Only if `N_draw` is > 0.

        HMF_draws : ndarray, optional
            HMF realizations drawn from the emulator posterior distribution. The
            return values `HMF_mean` and `HMF_std` are computed from these. Only
            if `N_draw` is > 0 and if `return_draws` is True.
        """

        # Validate and normalize requested cosmology
        requested_cosmology_normed = self.__normalize_params(requested_cosmology)

        output = {'Units': "log10_M is log10(Mass in [Msun/h]), HMFs are given in dN/dlnM [(h/Mpc)^3]"}
        log10_M_full = np.linspace(13, 16, 3001)
        for i in range(len(self.z_arr)):
            output[self.z_arr[i]] = {'redshift': self.z_arr[i],
                                     'log10_M': log10_M_full[:len(self.__PCA_means[i])],}

            # Call the GP
            wstar, wstar_covmat = self.__GPreg[i].predict(requested_cosmology_normed)
            # De-standardize to GP input
            PC_weight = wstar * self.__GP_std[i] + self.__GP_means[i]
            # PCA transform
            output[self.z_arr[i]]['HMF'] = np.exp(np.dot(PC_weight, self.__PCA_transform[i]) + self.__PCA_means[i])

            # Draw parameter realizations
            if N_draw>0:
                wstar_draws = np.random.multivariate_normal(wstar, wstar_covmat, N_draw)
                PC_weight_draws = wstar_draws * self.__GP_std[i] + self.__GP_means[i]
                HMF_draws = np.exp(np.dot(PC_weight_draws, self.__PCA_transform[i]) + self.__PCA_means[i])
                # Replace infinites with nan to be able to get mean and std
                idx = [np.all(np.isfinite(HMF_draws[j])) for j in range(N_draw)]
                output[self.z_arr[i]]['HMF_draws'] = HMF_draws[idx]
                # Compute statistics
                output[self.z_arr[i]]['HMF_mean'] = np.mean(output[self.z_arr[i]]['HMF_draws'], axis=0)
                output[self.z_arr[i]]['HMF_std'] = np.std(output[self.z_arr[i]]['HMF_draws']/output[self.z_arr[i]]['HMF_mean'], axis=0)
                if not return_draws:
                    del output[self.z_arr[i]]['HMF_draws']

        return output


    def validate_params(self, cosmo_dict):
        """Validate that a given input cosmology dictionary is complete and
        within the bounds of the Mira-Titan Universe design. Note that this
        function is only there for user convenience and is not called by the
        emulator code (which will raise an error and explain its cause).

        Parameters
        ----------
        cosmo_dict : dictionary
            The set of cosmology parameters to validate.

        Returns
        -------
        valid : bool
            Whether the provided dictionary is valid or not.
        """
        # Validate joint limit in w_0, w_a, then add w_b
        for param in ['w_0', 'w_a']:
            if param not in cosmo_dict.keys():
                return False
        if cosmo_dict['w_a'] > -cosmo_dict['w_0']:
            return False
        cosmo_dict['w_b'] = (-cosmo_dict['w_0'] - cosmo_dict['w_a'])**.25
        # Check all keys and parameter ranges
        for param in self.param_limits.keys():
            if param not in cosmo_dict.keys():
                return False
            if cosmo_dict[param]<self.param_limits[param][0]:
                return False
            if cosmo_dict[param]>self.param_limits[param][1]:
                return False
        return True


    def __normalize_params(self, cosmo_dict):
        """Check that a given input cosmology dictionary is complete and within
        the bounds of the Mira-Titan Universe design and return the normalized
        cosmological parameter array."""
        # Validate joint limit in w_0, w_a, then add w_b
        for param in ['w_0', 'w_a']:
            if param not in cosmo_dict.keys():
                raise KeyError("You did not provide %s"%param)
        if cosmo_dict['w_a'] > -cosmo_dict['w_0']:
            raise ValueError("w_0 + w_a must be <0. You have w_0 %.4f and w_a %.4f"%(cosmo_dict['w_0'], cosmo_dict['w_a']))
        cosmo_dict['w_b'] = (-cosmo_dict['w_0'] - cosmo_dict['w_a'])**.25
        # Check keys and parameter ranges
        for param in self.param_limits.keys():
            if param not in cosmo_dict.keys():
                raise KeyError("You did not provide %s"%param)
            if cosmo_dict[param]<self.param_limits[param][0]:
                raise ValueError("Parameter %s is %.4f but must be >= %.4f"%(param, cosmo_dict[param], self.param_limits[param][0]))
            if cosmo_dict[param]>self.param_limits[param][1]:
                raise ValueError("Parameter %s is %.4f but must be <= %.4f"%(param, cosmo_dict[param], self.param_limits[param][1]))
        # Normalize the parameters
        normed_p = np.empty(len(self.__param_names))
        for i,param in enumerate(self.__param_names):
            normed_p[i] = (cosmo_dict[param] - self.param_limits[param][0]) / (self.param_limits[param][1] - self.param_limits[param][0])

        return normed_p
